Add shellcheck

AliSoftware opened the pull request:
What it says on the tin.

Ensures the `pre-checkout` script follows code style and more importantly, doesn't contain accidental syntax/logic errors or fall into subtle bash traps.

### To Test

 - I expect the first CI runs to find and report lint warnings (at least a warning `SC2086` on line 22 like I have when running it locally), which would be a good check to ensure that lint step works and the CI check fails as expected.
 - Once we confirmed the warning is reported, we can fix it, push the fix, and validate the build and CI check passes

AliSoftware committed:
Add shellcheck CI step

AliSoftware commented:
I'm a bit confused why the CI fails to clone the repo, while I think I've followed all the relevant steps 🤔 
 - Created a ssh keypair
 - Added it to mobile-secrets (under `git-s3-cache` subdirectory) and pushed it
 - Confirmed via both #platform-alerts notification, and via `aws s3 ls`, that the key had been deployed to S3
 - Added the key [as a Deploy Key in this repos' settings GitHub](https://github.com/Automattic/git-s3-cache-buildkite-plugin/settings/keys)
 - Restarted the buildkite job… but still got the `Permission denied (publickey).` / `fatal: Could not read from remote repository.` errors 🤔 

The Buildkite logs seems to confirms that it found a private key in S3 and added it to the ssh agent, so I'm not sure what's wrong there 🤔 I even verified that I didn't copied the wrong Deploy Key by mistake, by running `ssh-keygen -l -f ~/.mobile-secrets/CI/secrets/git-s3-cache/private_ssh_key` and comparing its SHA signature with the Deploy Key, but all seems to match here 🤔  cc @jkmassel @mokagio any idea?

AliSoftware commented:
Added `GIT_SSH_COMMAND: "ssh -vvv"` to the `pipeline-upload` step's `env` in [Buildkite' UI](https://buildkite.com/automattic/git-s3-cache/steps), and got [the following logs](https://buildkite.com/automattic/git-s3-cache/builds/7#01820149-c9df-4356-a0fc-58fafb011fc6):

```
  | [2022-07-15T09:57:30Z] debug1: Offering ECDSA public key: mobile@automattic.com
  | [2022-07-15T09:57:30Z] debug3: send_pubkey_test
  | [2022-07-15T09:57:30Z] debug3: send packet: type 50
  | [2022-07-15T09:57:30Z] debug2: we sent a publickey packet, wait for reply
  | [2022-07-15T09:57:30Z] debug3: receive packet: type 51
```
(And according to Google, type 51 is `SSH_MSG_USERAUTH_FAILURE`)
Still doesn't help much at understanding _why_ it's refused 🤔 

<details><summary>More verbose extract of SSH logs</summary>

```
> Downloading secrets from  a8c-apps-ci-secrets | 0s
  | [2022-07-15T09:57:29Z] Checking S3 for SSH keys:
  | [2022-07-15T09:57:29Z] - git-s3-cache/private_ssh_key
  | [2022-07-15T09:57:29Z] - git-s3-cache/id_rsa_github
  | [2022-07-15T09:57:29Z] - private_ssh_key
  | [2022-07-15T09:57:29Z] - id_rsa_github
  | [2022-07-15T09:57:29Z] Checking S3 for environment files:
  | [2022-07-15T09:57:29Z] - env
  | [2022-07-15T09:57:29Z] - environment
  | [2022-07-15T09:57:29Z] - git-s3-cache/env
  | [2022-07-15T09:57:29Z] - git-s3-cache/environment
  | [2022-07-15T09:57:29Z] Checking S3 for git credentials:
  | [2022-07-15T09:57:29Z] - git-credentials
  | [2022-07-15T09:57:29Z] - git-s3-cache/git-credentials
  | [2022-07-15T09:57:29Z] Started ephemeral ssh-agent (pid 3346)
  | [2022-07-15T09:57:29Z] Loading a8c-apps-ci-secrets/git-s3-cache/private_ssh_key (748 bytes) into ssh-agent (pid 3346)
  | [2022-07-15T09:57:29Z] Loading a8c-apps-ci-secrets/env (679 bytes) of env

> Preparing working directory | 5s
  | [2022-07-15T09:57:29Z] # Creating "/var/lib/buildkite-agent/builds/ci-builders-i-066b05c9c22879806-1/automattic/git-s3-cache"
  | [2022-07-15T09:57:29Z] $ cd /var/lib/buildkite-agent/builds/ci-builders-i-066b05c9c22879806-1/automattic/git-s3-cache
  | [2022-07-15T09:57:29Z] # Added host "github.com" to known hosts at "/var/lib/buildkite-agent/.ssh/known_hosts"
  | [2022-07-15T09:57:29Z] $ git clone -v -- github.com:Automattic/git-s3-cache-buildkite-plugin.git .
  | [2022-07-15T09:57:29Z] Cloning into '.'...
  | [2022-07-15T09:57:29Z] OpenSSH_7.4p1, OpenSSL 1.0.2k-fips  26 Jan 2017
…
  | [2022-07-15T09:57:29Z] debug1: Authenticating to github.com:22 as 'buildkite-agent'
  | [2022-07-15T09:57:29Z] debug3: hostkeys_foreach: reading file "/var/lib/buildkite-agent/.ssh/known_hosts"
  | [2022-07-15T09:57:29Z] debug3: record_hostkey: found key type ECDSA in file /var/lib/buildkite-agent/.ssh/known_hosts:1
  | [2022-07-15T09:57:29Z] debug3: record_hostkey: found key type RSA in file /var/lib/buildkite-agent/.ssh/known_hosts:2
  | [2022-07-15T09:57:29Z] debug3: record_hostkey: found key type ED25519 in file /var/lib/buildkite-agent/.ssh/known_hosts:3
  | [2022-07-15T09:57:29Z] debug3: load_hostkeys: loaded 3 keys from github.com
…
  | [2022-07-15T09:57:29Z] debug1: Server host key: ecdsa-sha2-nistp256 SHA256:p2QAMXNIC1TJYWeIOttrVc98/R1BUFWu3/LiyKgUfQM
  | [2022-07-15T09:57:29Z] debug3: hostkeys_foreach: reading file "/var/lib/buildkite-agent/.ssh/known_hosts"
  | [2022-07-15T09:57:29Z] debug3: record_hostkey: found key type ECDSA in file /var/lib/buildkite-agent/.ssh/known_hosts:1
  | [2022-07-15T09:57:29Z] debug3: record_hostkey: found key type RSA in file /var/lib/buildkite-agent/.ssh/known_hosts:2
  | [2022-07-15T09:57:29Z] debug3: record_hostkey: found key type ED25519 in file /var/lib/buildkite-agent/.ssh/known_hosts:3
  | [2022-07-15T09:57:29Z] debug3: load_hostkeys: loaded 3 keys from github.com
  | [2022-07-15T09:57:29Z] debug3: hostkeys_foreach: reading file "/var/lib/buildkite-agent/.ssh/known_hosts"
  | [2022-07-15T09:57:29Z] debug1: Host 'github.com' is known and matches the ECDSA host key.
  | [2022-07-15T09:57:29Z] debug1: Found key in /var/lib/buildkite-agent/.ssh/known_hosts:1
  | [2022-07-15T09:57:29Z] Warning: Permanently added the ECDSA host key for IP address '140.82.112.4' to the list of known hosts.
…
  | [2022-07-15T09:57:30Z] debug2: key: mobile@automattic.com (0x55e0076c56b0), agent
  | [2022-07-15T09:57:30Z] debug2: key: /var/lib/buildkite-agent/.ssh/id_rsa ((nil))
  | [2022-07-15T09:57:30Z] debug2: key: /var/lib/buildkite-agent/.ssh/id_dsa ((nil))
  | [2022-07-15T09:57:30Z] debug2: key: /var/lib/buildkite-agent/.ssh/id_ecdsa ((nil))
  | [2022-07-15T09:57:30Z] debug2: key: /var/lib/buildkite-agent/.ssh/id_ed25519 ((nil))
…
  | [2022-07-15T09:57:30Z] debug1: Authentications that can continue: publickey
  | [2022-07-15T09:57:30Z] debug3: start over, passed a different list publickey
  | [2022-07-15T09:57:30Z] debug3: preferred gssapi-keyex,gssapi-with-mic,publickey,keyboard-interactive,password
  | [2022-07-15T09:57:30Z] debug3: authmethod_lookup publickey
  | [2022-07-15T09:57:30Z] debug3: remaining preferred: keyboard-interactive,password
  | [2022-07-15T09:57:30Z] debug3: authmethod_is_enabled publickey
  | [2022-07-15T09:57:30Z] debug1: Next authentication method: publickey
  | [2022-07-15T09:57:30Z] debug1: Offering ECDSA public key: mobile@automattic.com
  | [2022-07-15T09:57:30Z] debug3: send_pubkey_test
  | [2022-07-15T09:57:30Z] debug3: send packet: type 50
  | [2022-07-15T09:57:30Z] debug2: we sent a publickey packet, wait for reply
  | [2022-07-15T09:57:30Z] debug3: receive packet: type 51
  | [2022-07-15T09:57:30Z] debug1: Authentications that can continue: publickey
  | [2022-07-15T09:57:30Z] debug1: Trying private key: /var/lib/buildkite-agent/.ssh/id_rsa
  | [2022-07-15T09:57:30Z] debug3: no such identity: /var/lib/buildkite-agent/.ssh/id_rsa: No such file or directory
  | [2022-07-15T09:57:30Z] debug1: Trying private key: /var/lib/buildkite-agent/.ssh/id_dsa
  | [2022-07-15T09:57:30Z] debug3: no such identity: /var/lib/buildkite-agent/.ssh/id_dsa: No such file or directory
  | [2022-07-15T09:57:30Z] debug1: Trying private key: /var/lib/buildkite-agent/.ssh/id_ecdsa
  | [2022-07-15T09:57:30Z] debug3: no such identity: /var/lib/buildkite-agent/.ssh/id_ecdsa: No such file or directory
  | [2022-07-15T09:57:30Z] debug1: Trying private key: /var/lib/buildkite-agent/.ssh/id_ed25519
  | [2022-07-15T09:57:30Z] debug3: no such identity: /var/lib/buildkite-agent/.ssh/id_ed25519: No such file or directory
  | [2022-07-15T09:57:30Z] debug2: we did not send a packet, disable method
  | [2022-07-15T09:57:30Z] debug1: No more authentication methods to try.
  | [2022-07-15T09:57:30Z] Permission denied (publickey).
```
<summary>

AliSoftware commented:
Tried to regenerate a new keypair and update it both in mobile-secrets and in GitHub Settings' Deploy Keys, just in case (and because I couldn't see what else might be wrong), and it [still didn't solve the issue](https://buildkite.com/automattic/git-s3-cache/builds/11#0182029d-89c0-41c2-9358-bcd5fc819dff) apparently 😢 

So at that point I'll just let someone else give it a try at figuring out what's wrong, while I go AFK 🙃 

---

🧹  PS: Note to whomever might take over trying to fix that (@jkmassel ?), some debug things not to forget to clean up:
 - [x] Remove the extra [Deploy Key from the GitHub Settings](https://github.com/Automattic/git-s3-cache-buildkite-plugin/settings/keys), to only keep the working one
 - [x] Remove the `GIT_SSH_COMMAND` env var [from Buildkite's pipeline upload step](https://buildkite.com/automattic/git-s3-cache/steps)

mokagio commented:
I looked into this a bit, too. 

I applied the same trick to the [`bash-cache-github-plugin` pipeline](https://buildkite.com/automattic/bash-cache-buildkite-plugin/builds/214#01820f92-0dd4-496c-9d1c-9471efd72252). The build there succeeds, and the logs are, as far as I can see, the same, with the exception that the SSH check works:

```
  | [2022-07-18T04:30:14Z] debug1: Offering RSA public key: jkmassel@Jeremys-iMac.local
  | [2022-07-18T04:30:14Z] debug3: send_pubkey_test
  | [2022-07-18T04:30:14Z] debug3: send packet: type 50
  | [2022-07-18T04:30:14Z] debug2: we sent a publickey packet, wait for reply
  | [2022-07-18T04:30:14Z] debug3: receive packet: type 60
  | [2022-07-18T04:30:14Z] debug1: Server accepts key: pkalg rsa-sha2-512 blen 407
```

The key is of course different, but that's to be expected. Also, the fact that Olivier regenerated the key and tested it locally points to the issue not being with the key 🤔 

I noticed the key used in that repo uses `ssh-rsa` while the one for this repo `ecdsa-sha2-nistp521`. However, I checked other repos and there's plenty of `ecdsa-sha2-nistp521` that work.

mokagio commented:
I tried to rename the Buildkite pipeline and matching secrets folder in `mobile-secrets` to have the same name as the repo.

I didn't expect this to make a difference, because I could see that our automation successfully loaded the SSH key, but I thought it was worth a shot.

Unsurprisingly, it didn't make any difference.

However, I think we should stick with the updated, consistent name. What do you think?

mokagio commented:
I created a new pipeline for this repo and everything works there. 

🤷‍♂️

mokagio commented:
The update pipeline now fails _correctly_. That is, it picks up a ShellCheck error 🎉 

<img width="1157" alt="image" src="https://user-images.githubusercontent.com/1218433/179453463-8902e536-f0d0-422e-9885-28d97bccf534.png">

mokagio committed:
Address ShellCheck violation in `pre-checkout`

mokagio commented:
> The update pipeline now fails correctly. That is, it picks up a ShellCheck error 🎉

I addressed the violation, but I want to verify the behavior further before cherry-picking the fix and merging.

See: 

- https://github.com/Automattic/git-s3-cache-buildkite-plugin/commit/1da86c830e0730ae621b229d0155e88aaef50ad4
- #9

mokagio reviewed (approved):
Finally got to the bottom of this 🎉 

I verified that [my ShellCheck fix](https://github.com/Automattic/git-s3-cache-buildkite-plugin/pull/7/commits/d68c05359f67c3938d94a62f4ecddeeab5e7078e) also doesn't introduce a regression in the code.

Turns out, that `grep` call hasn't worked in a while, if ever 😱 

I have a fix for it (see behavior of [this WordPress iOS build](https://buildkite.com/automattic/wordpress-ios/builds/8933#01821461-5cda-492c-804f-867abe4965b6)) but I'll open a dedicated PR for it, for clarity.

In the meantime, I'll take over merging this one.

AliSoftware commented:
@mokagio Thanks for looking into this and getting to the bottom of it!
Even if the reason for the initial failure is still a mystery, at least we ultimately got it working, which is the most important 😄 

Closing the loop, I just addressed [the cleanup actions I mentioned above](https://github.com/Automattic/git-s3-cache-buildkite-plugin/pull/7#issuecomment-1185692253), aka:
 - Deleted all the [Deploy Keys](https://github.com/Automattic/git-s3-cache-buildkite-plugin/settings/keys) we used for debugging, only keeping the one we actually used
 - Removed the `GIT_SSH_COMMAND` env var [from the pipeline](https://buildkite.com/automattic/git-s3-cache/steps)

Then I opened https://github.com/Automattic/git-s3-cache-buildkite-plugin/pull/12 to check that everything still worked after that Deploy Keys cleanup ✅
